model_signing/manifest/manifest.py

```diff
@@ -37,3 +73,113 @@ class DigestManifest(Manifest):
     """A manifest that is just a hash."""
 
     digest: hashing.Digest
+
+
+class ItemizedManifest(Manifest):
+    """A detailed manifest, recording integrity of every model component."""
+
+    pass
+
+
+class ManifestItem(metaclass=abc.ABCMeta):
+    """An object of a model that can be stored in an `ItemizedManifest`.
+
+    For example, this could be a file, or a file shard. All file paths are
+    relative to the model root, to allow moving or renaming the model, without
+    invalidating the signature.
+
+    The integrity of each `ManifestItem` can be verified individually and in
+    parallel. If the item is backed by a file, we recompute the hash of the
+    portion of the file that represents this item.
+    """
+
+    pass
+
+
+@dataclasses.dataclass
+class FileManifestItem(ManifestItem):
+    """A manifest item that records a filename path together with its digest.
+
+    Note that the path component is a `pathlib.PurePath`, relative to the model.
+    """
+
+    path: pathlib.PurePath
+    digest: hashing.Digest
+
+    def __init__(self, *, path: pathlib.PurePath, digest: hashing.Digest):
+        """Builds a manifest item pairing a file with its digest.
+
+        Args:
+            path: The path to the file, relative to the model root.
+            digest: The digest of the file.
+        """
+        # Note: we need to force a PosixPath to canonicalize the manifest
+        self.path = pathlib.PurePosixPath(path)
+        self.digest = digest
+
+
+class FileLevelManifest(ItemizedManifest):
+    """A detailed manifest, recording integrity of every model file."""
+
+    def __init__(self, items: Iterable[FileManifestItem]):
+        """Builds an itemized manifest from a collection of files.
+
+        Rather than recording the items in a list, we use a dictionary, to allow
+        efficient updates and retrieval of digests.
+        """
+        self._digest_info = {item.path: item.digest for item in items}
+
+    def __eq__(self, other: Self):
+        return self._digest_info == other._digest_info
+
+
+@dataclasses.dataclass
+class ShardedFileManifestItem(ManifestItem):
+    """A manifest item that records a file shard together with its digest."""
+
+    path: pathlib.PurePath
+    start: int
+    end: int
+    digest: hashing.Digest
+
+    def __init__(
+        self,
+        *,
+        path: pathlib.PurePath,
+        start: int,
+        end: int,
+        digest: hashing.Digest
+    ):
+        """Builds a manifest item pairing a file with its digest.
+
+        Args:
+            path: The path to the file, relative to the model root.
+            start: The start offset of the shard.
+            end: The end offset of the shard.
+            digest: The digest of the file.
+        """
```

Comment:
may be a minor copy/paste here in terms of what the digest applies to.

> Builds a manifest item pairing a **file** with its digest.

> digest: The digest of the **file**.

I believe these should both be "file shard"

Reply:
Nice catch, fixed!